lib/ansible/plugins/strategy/__init__.py

```diff
@@ -652,20 +652,28 @@ def _process_pending_results(self, iterator, one_pass=False, max_passes=None):
                         # only ensure that notified handlers exist, if so save the notifications for when
                         # handlers are actually flushed so the last defined handlers are exexcuted,
                         # otherwise depending on the setting either error or warn
+                        host_state = iterator.get_state_for_host(original_host.name)
                         for notification in result_item['_ansible_notify']:
-                            if any(self.search_handlers_by_notification(notification, iterator)):
-                                iterator.add_notification(original_host.name, notification)
-                                display.vv(f"Notification for handler {notification} has been saved.")
-                                continue
-
-                            msg = (
-                                f"The requested handler '{notification}' was not found in either the main handlers"
-                                " list nor in the listening handlers list"
-                            )
-                            if C.ERROR_ON_MISSING_HANDLER:
-                                raise AnsibleError(msg)
+                            for handler in self.search_handlers_by_notification(notification, iterator):
+                                if host_state.run_state == IteratingStates.HANDLERS:
+                                    # we're currently iterating handlers, so we need to expand this now
+                                    if handler.notify_host(original_host):
+                                        # NOTE even with notifications deduplicated this can still happen in case of handlers being
+                                        # notified multiple times using different names, like role name or fqcn
+                                        self._tqm.send_callback('v2_playbook_on_notify', handler, original_host.name)
```

Comment:
Just noticed this:
```suggestion
                                        self._tqm.send_callback('v2_playbook_on_notify', handler, original_host)
```

We historically passed a host object to `v2_playbook_on_notify`, not a name and we already pass a host object in `execute_meta`: 
https://github.com/ansible/ansible/blob/86664c4cce1e3a44572df72e7f8280c524e12eb2/lib/ansible/plugins/strategy/__init__.py#L961

---
Although it does not matter in the current code base as whatever is passed is printed as a string:
https://github.com/ansible/ansible/blob/058a69e6b037c87229d1d28742458e90fc1522cd/lib/ansible/plugins/callback/default.py#L409

so passing a host object is effectively turned into `host.name` through "%s" -> `__str__` -> `host.get_name()` -> `host.name`. But we should still pass a host object as that is probably what was originally intended, type hints will help here :-)

Reply:
Will be fixed in next push.